src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+) -> Result<bool, ScriptError> {
+    let bn_zero = ScriptNum::from(0);
+    let bn_one = ScriptNum::from(1);
+    let vch_false: ValType = vec![];
+    let vch_true: ValType = vec![1];
+
+    // There's a limit on how large scripts can be.
+    if script.0.len() > MAX_SCRIPT_SIZE {
+        return set_error(ScriptError::ScriptSize);
+    }
+
+    let mut pc = script.0;
+    let mut vch_push_value = vec![];
+
+    // We keep track of how many operations have executed so far to prevent
+    // expensive-to-verify scripts
+    let mut op_count: u8 = 0;
+    let require_minimal = flags.contains(VerificationFlags::MinimalData);
+
+    // This keeps track of the conditional flags at each nesting level
+    // during execution. If we're in a branch of execution where *any*
+    // of these conditionals are false, we ignore opcodes unless those
+    // opcodes direct control flow (OP_IF, OP_ELSE, etc.).
+    let mut vexec: Stack<bool> = Stack(vec![]);
+
+    let mut altstack: Stack<Vec<u8>> = Stack(vec![]);
+
+    // Main execution loop
+    while !pc.is_empty() {
+        // Are we in an executing branch of the script?
+        let exec = vexec.iter().all(|value| *value);
+
+        //
+        // Read instruction
+        //
+        let opcode = Script::get_op2(&mut pc, &mut vch_push_value)?;
+        if vch_push_value.len() > MAX_SCRIPT_ELEMENT_SIZE {
+            return set_error(ScriptError::PushSize);
+        }
+
+        match opcode {
+            Opcode::PushValue(pv) => {
+                if exec {
+                    match pv {
+                        //
+                        // Push value
+                        //
+                        OP_1NEGATE | OP_1 | OP_2 | OP_3 | OP_4 | OP_5 | OP_6 | OP_7 | OP_8
+                        | OP_9 | OP_10 | OP_11 | OP_12 | OP_13 | OP_14 | OP_15 | OP_16 => {
+                            // ( -- value)
+                            let bn = ScriptNum::from(u8::from(pv)) - u8::from(OP_RESERVED).into();
+                            stack.push_back(bn.getvch());
+                            // The result of these opcodes should always be the minimal way to push the data
+                            // they push, so no need for a CheckMinimalPush here.
+                        }
+                        _ => {
+                            if pv <= OP_PUSHDATA4 {
+                                if require_minimal && !check_minimal_push(&vch_push_value, pv) {
+                                    return set_error(ScriptError::MinimalData);
+                                }
+                                stack.push_back(vch_push_value.clone());
+                            } else {
+                                return set_error(ScriptError::BadOpcode);
+                            }
+                        }
+                    }
+                }
+            }
+            Opcode::Operation(op) => {
+                // Note how OP_RESERVED does not count towards the opcode limit.
```

Comment:
This comment is from the C++ code but I do not understand it. In C++, any `opcode > OP_16` *does* count toward the opcode limit, and that includes `OP_RESERVED` which is `0x50`.